src/magic_folder/_schema.py

```diff
@@ -0,0 +1,148 @@
+# Copyright 2020 Least Authority TFA GmbH
+# See COPYING for details.
+
+"""
+This module defines the database schema used by the model interface.
+
+:var int MAXIMUM_UPGRADES: The maximum number of upgrades that are allowed in
+    a single schema.
+"""
+
+from __future__ import (
+    unicode_literals,
+    absolute_import,
+)
+
+import attr
+
+_CREATE_VERSION = (
+    """
+    CREATE TABLE IF NOT EXISTS [schema-version] AS SELECT 0 AS [version]
+    """
+)
+
+_INCREMENT_VERSION = (
+    """
+    UPDATE [schema-version]
+    SET [version] = [version] + 1
+    """
+)
+
+_READ_VERSION = (
+    """
+    SELECT [version] FROM [schema-version]
+    """
+)
+
+MAXIMUM_UPGRADES = 1000
+
+@attr.s(frozen=True)
+class DatabaseSchemaTooNew(Exception):
+    """
+    The schema in the database is newer than the Python schema representation.
+
+    This version of the software cannot use this version of the database.
+    """
+    software_version = attr.ib()
+    database_version = attr.ib()
+
```

Comment:
It would be better if the latter happened, yea.  However, that Exception/attrs behavior sucks and if there's even a chance of anyone having to look at the str() of this exception, intentionally or unintentionally, I'd rather that be something other than "" so I guess I'll change this.